llvm/lib/Transforms/IPO/SampleProfile.cpp

```diff
@@ -2184,6 +2204,61 @@ bool SampleProfileLoader::doInitialization(Module &M,
   return true;
 }
 
+// Note that this is a module-level check. Even if one module is errored out,
+// the entire build will be errored out. However, the user could make big
+// changes to functions in single module but those changes might not be
+// performance significant to the whole binary. Therefore, we use a conservative
+// approach to make sure we only error out if it globally impacts the binary
+// performance. To achieve this, we use heuristics to select a reasonable
+// big set of functions that are supposed to be globally performance
+// significant, only compute and check the mismatch within those functions. The
+// function selection is based on two criteria: 1) The function is "hot" enough,
+// which is tuned by a hotness-based flag(ChecksumMismatchFuncHotBlockSkip). 2)
+// The num of function is large enough which is tuned by the
+// ChecksumMismatchNumFuncSkip flag.
+bool SampleProfileLoader::errorIfHighChecksumMismatch(
+    Module &M, ProfileSummaryInfo *PSI, const SampleProfileMap &Profiles) {
+  assert(FunctionSamples::ProfileIsProbeBased &&
+         "Only support for probe-based profile");
+  uint64_t TotalSelectedFunc = 0;
+  uint64_t NumMismatchedFunc = 0;
+  for (const auto &I : Profiles) {
+    const auto &FS = I.second;
+    const auto *FuncDesc = ProbeManager->getDesc(FS.getGUID());
+    if (!FuncDesc)
+      continue;
+
+    // We want to select a set of functions that are globally performance
+    // significant, in other words, if those functions profiles are
+    // checksum-mismatched and dropped, the whole binary will likely be
+    // impacted, so here we use a hotness-based threshold to control the
+    // selection.
+    if (FS.getTotalSamples() <
+        ChecksumMismatchFuncHotBlockSkip * PSI->getOrCompHotCountThreshold())
```

Comment:
> 1. The multiplier doesn't really has to do with "block" of the function, so the name can be confusing?
> 2. What do you think of using `isFunctionHotInCallGraphNthPercentile` or `isHotCountNthPercentile` with a percentile cutoff? We can tun the percentile to get what we need.

Let me try to clarify my initial thought: IIUC,  all the "HotCount" stuffs in PSI are "block-level"(or body-sample level) concept,  the `isFunctionHotXXX` function are used to check the **entry_count** of the function which is also "block-level"(the first block), however what I used here is the `getTotalSamples()` which I intend to include all the callee/inlinee samples(it could be the entry count is cold, but it does a lot of hot inlinings), as total_samples is kind of sum of all the block counts of this function, so I used a multiplier which I want to mean how many blocks on average is hot in this function.   

So if we pass  a sum of blocks(`getTotalSamples`) to this `isFunctionHotInCallGraphNthPercentile ` or `isHotCountNthPercentile ` function, might still be confusing? 

I feet we actually don't have a function-level thing for this hotness check. so I was also thinking to add a new function to PSI, like:(but no sure if it's generally useful)
```
int getNumHotBlocks(const FunctionSamples &FS) {
   int num = 0; 
   for(samples in all body samples) 
      if(samples > isHotCount())
         num++;
    return num;
}
```
Any thought? thanks!

Reply:
> the isFunctionHotXXX function are used to check the entry_count of the function which is also "block-level"(the first block), however what I used here is the getTotalSamples() which I intend to include all the callee/inlinee samples(it could be the entry count is cold, but it does a lot of hot inlinings)

So there are different ways to decide whether a function is hot. Here we want to select a few hot functions and see how many of them has checksum mismatch. Checking on total samples is one way, and is different from existing isFunctionHotXXX. But do we really need to add another way of deciding whether a function is hot? I don't know the answer, but maybe use `isFunctionHotInCallGraphNthPercentile` to decide whether function is hot for our purpose is ok too? 

I agree that passing total samples into `isHotCountNthPercentile` maybe not be a good idea since the latter expects block counts.

Reply:
I see, sounds good to use `total_samples` to decide if the function is hot. 

I took a look at `isFunctionHotInCallGraphNthPercentile `, looks like the function's parameter is `Function` (`isFunctionHotOrColdInCallGraphNthPercentile(int PercentileCutoff,
                                                   const FuncT *F,
                                                   BFIT &FI)`) https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/Analysis/ProfileSummaryInfo.h#L285 , and it does looks for the annotations on the IR to check the hotness, however, here our function is before the sample annotation, so we can't directly use this function. 
 Actually I found that in `isFunctionHotOrColdInCallGraphNthPercentile` https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/Analysis/ProfileSummaryInfo.h#L285 , it also uses `isHotCountNthPercentile` for `TotalCallCount` which seems also not block-level counts, probably right now we don't have an existing function for total_samples, so maybe to make it simple, we can just use `isHotCountNthPercentile` for total_samples? (Or maybe add a overload function like `isFunctionHotInCallGraphNthPercentile(... , total_samples) ` to accept total_samples, but I think it's really just a wrapper of `isHotCountNthPercentile`)

Reply:
Digging more, I found that the `callsiteIsHot `https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Utils/SampleProfileLoaderBaseUtil.cpp#L64  also uses total_samples to decide hot function.. maybe we can use it

Reply:
Sounds good to just use `isHotCountNthPercentile` against total samples for now. 

> Maybe a different topic, but I'm surprised why it uses total_samples to make inlining decision..

Yes, this is not ideal. We intentionally switched away from that for CSSPGO. So if you check, the path that uses `callsiteIsHot` should be AutoFDO only.

Reply:
> Sounds good to just use `isHotCountNthPercentile` against total samples for now.
> 
> > Maybe a different topic, but I'm surprised why it uses total_samples to make inlining decision..
> 
> Yes, this is not ideal. We intentionally switched away from that for CSSPGO. So if you check, the path that uses `callsiteIsHot` should be AutoFDO only.

Good to know, checked that the usages of `callsiteIsHot` are indeed from AutoFDO path.